### PR TITLE
Add alluxio StacksServlet to display stacks on web ui page

### PR DESCRIPTION
**Pull request by maobaolong**
You can use `/stacks` to show the stack info in the web ui page for  alluxio master, worker, jobmaster, jobworker, proxy.

image

### CODE REVIEW

**Comment by apc999** [comment]
this is quite useful. @maobaolong  can you add documentation on this , maybe on 
https://github.com/Alluxio/alluxio/blob/master/docs/en/operation/Web-Interface.md ?

File: core/server/common/src/main/java/alluxio/ReflectionUtils.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+  private static String getTaskName(long id, String name) {
+    if (name == null) {
+      return Long.toString(id);
+    }
+    return id + " (" + name + ")";
+  }
+
+  /**
+   * Print all of the thread's information and stack traces.
```
Comment (apc999) [inline core/server/common/src/main/java/alluxio/ReflectionUtils.java:39]: nit, alluxio convention
```suggestion
   * Prints the information and stack traces of all threads.
```

File: core/server/common/src/main/java/alluxio/ReflectionUtils.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    stream.flush();
+  }
+
+  private static long sPreviousLogTime = 0;
+
+  /**
+   * Log the current thread stacks at INFO level.
+   * @param log the logger that logs the stack trace
+   * @param title a descriptive title for the call stacks
+   * @param minInterval the minimum time from the last
```
Comment (apc999) [inline core/server/common/src/main/java/alluxio/ReflectionUtils.java:88]: in unit of ms? please be specific

File: core/server/common/src/main/java/alluxio/ReflectionUtils.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      }
+      stream.println("  Stack:");
+      for (StackTraceElement frame : info.getStackTrace()) {
+        stream.println("    " + frame.toString());
+      }
+    }
+    stream.flush();
+  }
+
+  private static long sPreviousLogTime = 0;
```
Comment (apc999) [inline core/server/common/src/main/java/alluxio/ReflectionUtils.java:82]: `@GuardedBy("ReflectionUtils.class")`

File: core/server/common/src/main/java/alluxio/ReflectionUtils.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+import java.io.UnsupportedEncodingException;
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadInfo;
+import java.lang.management.ThreadMXBean;
+import java.nio.charset.Charset;
+
+/**
+ * General reflection utils.
+ */
+public final class ReflectionUtils {
```
Comment (apc999) [inline core/server/common/src/main/java/alluxio/ReflectionUtils.java:27]: Put methods to `alluxio.util.ThreadUtils` where we already have a few related util methods

File: core/server/common/src/main/java/alluxio/ReflectionUtils.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      } else if (state == Thread.State.BLOCKED) {
+        stream.println("  Blocked on " + info.getLockName());
+        stream.println("  Blocked by "
+            + getTaskName(info.getLockOwnerId(), info.getLockOwnerName()));
+      }
+      stream.println("  Stack:");
+      for (StackTraceElement frame : info.getStackTrace()) {
+        stream.println("    " + frame.toString());
+      }
+    }
```
Comment (apc999) [inline core/server/common/src/main/java/alluxio/ReflectionUtils.java:78]: refactor and leverge `ThreadUtils.logAllThreads` ?

**Review by apc999 — CHANGES_REQUESTED** [review]
Thanks for contributing this great feature!
Comments left

File: core/server/common/src/main/java/alluxio/ReflectionUtils.java (hunk 2 above)
Comment (maobaolong) [reply core/server/common/src/main/java/alluxio/ReflectionUtils.java]: Done

File: core/server/common/src/main/java/alluxio/ReflectionUtils.java (hunk 3 above)
Comment (maobaolong) [reply core/server/common/src/main/java/alluxio/ReflectionUtils.java]: Done

File: core/server/common/src/main/java/alluxio/ReflectionUtils.java (hunk 4 above)
Comment (maobaolong) [reply core/server/common/src/main/java/alluxio/ReflectionUtils.java]: Done

File: core/server/common/src/main/java/alluxio/ReflectionUtils.java (hunk 5 above)
Comment (maobaolong) [reply core/server/common/src/main/java/alluxio/ReflectionUtils.java]: Done

**Comment by maobaolong** [comment]
@apc999 Thank you for your review and suggestion, I did some refactor in the latest commit, PTAL.

File: core/common/src/main/java/alluxio/util/ThreadUtils.java [hunk 6]
```diff
@@ -92,5 +98,89 @@ public static String getThreadIdentifier(Thread thread) {
...
+  }
+
+  @GuardedBy("ThreadUtils.class")
+  private static long sPreviousLogTime = 0;
+
+  /**
+   * Log the current thread stacks at INFO level.
+   * @param log the logger that logs the stack trace
+   * @param title a descriptive title for the call stacks
+   * @param minInterval the minimum time from the last, unit is minute
```
Comment (apc999) [inline core/common/src/main/java/alluxio/util/ThreadUtils.java:159]: minute or second?

File: core/common/src/main/java/alluxio/util/ThreadUtils.java [hunk 7]
```diff
@@ -92,5 +98,89 @@ public static String getThreadIdentifier(Thread thread) {
...
+      stream.println("  Stack:");
+      for (StackTraceElement frame : info.getStackTrace()) {
+        stream.println("    " + frame.toString());
+      }
+    }
+    stream.flush();
+  }
+
+  @GuardedBy("ThreadUtils.class")
+  private static long sPreviousLogTime = 0;
```
Comment (apc999) [inline core/common/src/main/java/alluxio/util/ThreadUtils.java:153]: please document its unit, ms ?

**Review by apc999 — COMMENTED** [review]
last batch of comments left

File: core/common/src/main/java/alluxio/util/ThreadUtils.java (hunk 6 above)
Comment (maobaolong) [reply core/common/src/main/java/alluxio/util/ThreadUtils.java:159]: Sorry, unit is sec

File: core/common/src/main/java/alluxio/util/ThreadUtils.java (hunk 7 above)
Comment (maobaolong) [reply core/common/src/main/java/alluxio/util/ThreadUtils.java:153]: Done.

**Comment by maobaolong** [comment]
@apc999 Please take another look, thanks.

**Review by apc999 — APPROVED** [review]

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Pull request merged**